tests/unit/APNSResponseTest.php

```diff
@@ -19,4 +19,31 @@ public function testThatResponseParserReadsErrorResponse() {
 		$this->assertEquals( 400, $response->getStatusCode() );
 		$this->assertEquals( 'BadDeviceToken', $response->getErrorMessage() );
 	}
+
+	public function testUnrecoverableErrorEvaluation() {
+		$this->assertTrue( $this->makeAPNSResponseFor( 400 )->isUnrecoverableError() );
+		$this->assertTrue( $this->makeAPNSResponseFor( 403 )->isUnrecoverableError() );
+		$this->assertTrue( $this->makeAPNSResponseFor( 404 )->isUnrecoverableError() );
+		$this->assertTrue( $this->makeAPNSResponseFor( 405 )->isUnrecoverableError() );
+		$this->assertTrue( $this->makeAPNSResponseFor( 410 )->isUnrecoverableError() );
+		$this->assertTrue( $this->makeAPNSResponseFor( 413 )->isUnrecoverableError() );
+		$this->assertFalse( $this->makeAPNSResponseFor( 429 )->isUnrecoverableError() );
+	}
+
+	private function makeAPNSResponseFor( int $status_code ) {
+		return new APNSResponse(
+			$status_code,
+			$this->fakeAPNSHTTPFailureResponse( $status_code, 'not read here' ),
+			new APNSResponseMetrics()
+		);
+	}
```

Comment:
Because the only thing that matters in deciding whether an error response is unrecoverable is the status code, I hid all the rest in this method.

Reply:
Yep – this seems great. One nitpick I'd suggest is using a default parameter for something like this to make it easy to reuse this method for other future tests. Something like:

```php
	private function makeAPNSResponseFor( int $status_code, string $message = 'not read here' ) {
```

would do the same, but be a bit more flexible for later.

Reply:
Great suggestion! Thanks ✨ .